core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -73,103 +73,106 @@ protected void entryReplaced(long newAddress, long oldAddress) {
+    * @param address
+    */
+   private void removeNode(long address) {
+      boolean middleNode = true;
+      if (address == lastAddress) {
+         if (trace) {
+            log.tracef("Removed LRU node at end 0x%016x", address);
          }
-         if (lruNode == firstAddress) {
-            if (trace) {
-               log.tracef("Removing first LRU node at 0x%016x", lruNode);
-            }
-            long nextLRUNode = OffHeapLruNode.getNext(lruNode);
-            if (nextLRUNode != 0) {
-               OffHeapLruNode.setPrevious(nextLRUNode, 0);
-            }
-            firstAddress = nextLRUNode;
-            middleNode = false;
+         long previousLRUNode = OffHeapLruNode.getPrevious(address);
+         if (previousLRUNode != 0) {
+            OffHeapLruNode.setNext(previousLRUNode, 0);
          }
-         if (middleNode) {
-            if (trace) {
-               log.tracef("Removing middle LRU node at 0x%016x", lruNode);
-            }
-            // We are a middle pointer so both of these have to be non zero
-            long previousLRUNode = OffHeapLruNode.getPrevious(lruNode);
-            long nextLRUNode = OffHeapLruNode.getNext(lruNode);
-            assert previousLRUNode != 0;
-            assert nextLRUNode != 0;
-            OffHeapLruNode.setNext(previousLRUNode, nextLRUNode);
-            OffHeapLruNode.setPrevious(nextLRUNode, previousLRUNode);
+         lastAddress = previousLRUNode;
+         middleNode = false;
+      }
+      if (address == firstAddress) {
+         if (trace) {
+            log.tracef("Removing LRU node at beginning 0x%016x", address);
          }
-         allocator.deallocate(lruNode, OffHeapLruNode.getSize());
-      } finally {
-         lruLock.unlock();
+         long nextLRUNode = OffHeapLruNode.getNext(address);
+         if (nextLRUNode != 0) {
+            OffHeapLruNode.setPrevious(nextLRUNode, 0);
+         }
+         firstAddress = nextLRUNode;
+         middleNode = false;
+      }
+      if (middleNode) {
+         if (trace) {
+            log.tracef("Removing middle LRU node at 0x%016x", address);
+         }
+         // We are a middle pointer so both of these have to be non zero
+         long previousLRUNode = OffHeapLruNode.getPrevious(address);
+         long nextLRUNode = OffHeapLruNode.getNext(address);
+         assert previousLRUNode != 0;
+         assert nextLRUNode != 0;
+         OffHeapLruNode.setNext(previousLRUNode, nextLRUNode);
+         OffHeapLruNode.setPrevious(nextLRUNode, previousLRUNode);
       }
-      super.entryRemoved(removedAddress);
    }
 
    @Override
    protected void entryRetrieved(long entryAddress) {
       lruLock.lock();
       try {
-         long lruNode = offHeapEntryFactory.getLruNode(entryAddress);
          if (trace) {
-            log.tracef("Moving lruNode 0x%016x to the end which points at address 0x%016x", lruNode, entryAddress);
+            log.tracef("Moving node 0x%016x to the end", entryAddress);
```

Comment:
Here too, "Moving entry 0x%016x to the end of the LRU list"

Reply:
Sure